ENH add support to missing values in NMF

TomDLT opened the pull request:
Fixes #8447
Fixes #25229

Here I update NMF's multiplicative update solver to handle missing values (`np.nan`).
The solver simply doesn't take into account the loss at missing values, and optimizes the factorization on the rest of the data. This can be useful for imputation, recommendation (a bit out of scope for sklearn), or cross-validation.

~~**Not sure this is in scikit-learn scope though.**~~
___
The solver update is simple. Except for beta = 1, both denominator and numerator (in the multiplicative update) have either `X` or `W * H`, so we can mask them and sum the denominator and the numerator over valid elements. This is referred in the literature as "weighted NMF".

For the coordinate descent solver, the method uses the following factorization, to avoid recomputing `W * H` for each coordinate update of W: `(X - W * H) * H.T = X * H.T - W * (H * H.T)`
During the entire W-update, `X * H.T` and `H * H.T` do not change, so the optimization is fast.
But if we want to handle missing value, `H * H.T` changes for each line of W, which increases heavily the computations.
So I did not update the coordinate descent solver to handle missing values.

For the multiplicative update solver, the approximation `W * H` is used for the update of all coordinates, so we can compute it only once for each update of `W`. This is still more costly than `W * (H * H.T)` as used with beta = 2 without missing values, but this is already what we do for the other beta losses (!= 2).

I skipped the sparse case for now.
___
TODO:
- [ ] add more tests
- [ ] add some documentation and an example
- [ ] add to https://scikit-learn.org/stable/modules/impute.html

TomDLT committed:
add support to missing values in NMF

dburkhardt commented:
Is this still under development? I see multiple open issues referencing this functionality, but cannot tell if there have been any updates or if this is scheduled for a future release. I would really like to see this functionality!

jnothman commented:
I think it's open for someone to complete

TomDLT commented:
We talked briefly about this PR during previous sprint in June, and it was considered a bit out of scope of scikit-learn. Indeed, there is no other estimator that support missing values, and it will probably not work very well with the meta-estimators such as the Pipeline, GridSearchCV, ..

However, the code is working (though not reviewed), and could be moved to a separate repository properly referenced. I plan to do it at some point, but I am ok if someone wants to do it.

TomDLT committed:
Fix masked nonzero selection

Cristianasp commented:
Thanks ! Got this code and it is working with NaN values.

hongkahjun commented:
Used it with no problems!

jnothman commented:
I'd be interested in having an instance of estimating a latent representation to account for missing values as an alternative to imputation in a classification pipeline. Is that one of the applications of this? Could we add it to `plot_missing_values.py` and see how it compares?

TomDLT committed:
Fix a bug in NMF.transform with NaN values in X

TomDLT committed:
add ImputerNMF class

TomDLT commented:
Yes, imputation is one application of this, though one can also be interested in the NMF decomposition itself, and not the derived imputation.
To compare with `Imputer`, I added an `ImputerNMF` class which implement an imputation transformer with NMF, and added it one `plot_missing_values`.

The performances are not terrific though:
```
Score with the entire dataset = 0.56
Score without the samples containing missing values = 0.48
Score after imputation of the missing values (mean) = 0.57
Score after imputation of the missing values (NMF) = 0.51
```

jnothman commented:
I was not taking about imputation in the input source, but about deriving a
latent representation in which missing values are accounted for, and using
that to train.... I know similar is done with autoencoders.

TomDLT committed:
remove ImputerNMF

TomDLT committed:
keep the same memory for np.dot(W, H)

TomDLT commented:
Right, I removed the `ImputerNMF` which was not adding much...
About using directly the NMF representation in `plot_missing_values.py`, I changed the boston dataset into the olivetti face dataset, which better suits the factorization hypothesis of NMF.

The results are quite good:
```
Score with the entire dataset = 0.88
Score without the samples containing missing values = 0.68
Score after imputation of the missing values (mean) = 0.88
Score with NMF representation robust to missing values = 0.87
```

jnothman commented:
better, but not quite compelling!

TomDLT committed:
Fix incorrect ravel in non C-ordered array

leizing commented:
Hi, will this be only available in  v0.20.dev0?

TomDLT commented:
This functionality has been considered a bit out of scope of scikit-learn (at least for now).
To use it, you can fetch this code with:
```console
git fetch upstream pull/8474/head:nmf_missing
git checkout nmf_missing
```

leizing commented:
Many thanks. Agree that it may be out of score of scikit-learn, but the code would certainly be of many practical use. Thanks again.

leizing commented:
Hello, could you let me get access to the code (again)?  git fetch upstream pull xxx gives: 
` 'upstream' does not appear to be a git repository`

jnothman commented:
Replace upstream with https://github.com/scikit-learn/scikit-learn

TomDLT committed:
Merge branch 'master' into nmf_missing

sklearn-lgtm commented:
This pull request introduces 1 alert and fixes 2 - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-4718f89d058f18d16b7bc9c8fa8d65709da47648)

**new alerts:**

* 1 for Potentially uninitialized local variable

**fixed alerts:**

* 2 for Potentially uninitialized local variable

---

*Comment posted by [lgtm.com](https://lgtm.com)*

tnlin commented:
How is going? I want this features too. Is there anything I can help?

TomDLT committed:
Merge branch 'master' into nmf_missing

MelBeckerleg commented:
(originally posted https://github.com/scikit-learn/scikit-learn/issues/12233, redirected to comment here)
# problem

I am trying to get the `nmf_missing` branch to compile without success, [as per this suggestion](https://github.com/scikit-learn/scikit-learn/pull/8474#issuecomment-347520280).

## The compiler error:

On branch `nmf_missing` after issuing `make` at the command line it build largely fine until the following error (The first one I know of, although several warnings):

    error: Command "x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/include/python2.7 -c sklearn/neighbors/quad_tree.c -o build/temp.linux-x86_64-2.7/sklearn/neighbors/quad_tree.o -MMD -MF build/temp.linux-x86_64-2.7/sklearn/neighbors/quad_tree.o.d" failed with exit status 1
    Makefile:24: recipe for target 'inplace' failed


### How I think this can be solved(?)

It seems to come from trying to compile the following line in quad_tree.c (3582) `      printf(__pyx_k_QuadTree_found_a_duplicate);`. I have tried commenting out this line but whenever I type `make` it reverts the change. Hence I have tried tracking down what is producing either this file, or whatever is producing the compiler flags `-Wformat -Werror=format-security` which is causing the compiler to throw the error. However, trying to track the origin of these has proved too much and I am lost. 

#### Versions

python: 2.7.12, numpy 1.14.2, scikit-learn 0.20, cython 0.23.4

TomDLT committed:
Merge branch 'master' into nmf_missing

Conflicts:
	sklearn/utils/validation.py

TomDLT committed:
DOC add accept_nan in docstring

MelBeckerleg commented:
> I don't know what causes this problem.
> I have merged master into this branch, you may want to try again with the updated branch.
> You may also take a look at [the guide to install from source](http://scikit-learn.org/stable/developers/advanced_installation.html#building-from-source).

Tried again with the updated branch and updated numpy and now works. Thanks!

TomDLT committed:
Merge branch 'master' into nmf_missing

cmarmo commented:
@agramfort [comment](https://github.com/scikit-learn/scikit-learn/issues/8447#issuecomment-531564094) suggested to move this implementation to scikit-learn-contrib. I'm removing the "help wanted" label then. @TomDLT , please comment if there are any objections. Thanks.

glemaitre commented:
I think that this PR is more in the scope of `scikit-learn` nowadays. Indeed, our preprocessing methods are more lenient towards missing values and we have HGBDT that natively manages missing values. So the PR is quite relevant.

We would need to refactor the code with `main` then.

TomDLT committed:
Merge branch 'main' into nmf_missing

TomDLT committed:
extend to MinibatchNMF, apply black formatting

TomDLT committed:
fix merging conflict

TomDLT committed:
fix black version

TomDLT committed:
Merge branch 'main' into nmf_missing

TomDLT committed:
add changelog entry

TomDLT committed:
add allow_nan tag, add doc

TomDLT committed:
fix allow_nan tag

TomDLT committed:
fix check_array, simplify test diff

TomDLT committed:
black

TomDLT committed:
Merge branch 'main' into nmf_missing

TomDLT committed:
fix docstring, improve test

TomDLT committed:
black

TomDLT committed:
rework test_nmf_imputation

TomDLT committed:
Merge branch 'main' into nmf_missing

TomDLT committed:
Merge branch 'main' into nmf_missing

TomDLT committed:
Merge branch 'main' into nmf_missing

sammymans commented:
I know this is an old thread but has this been added to the main scikit-learn repo? or can this only be found in @TomDLT forked repo. How can i use the nmf_missing to handle the missing NaN values in a rating matrix for example

TomDLT commented:
Not merged yet, waiting for reviews.

In the meantime, you can checkout the code with:
```bash
git fetch https://github.com/scikit-learn/scikit-learn pull/8474/head:nmf_missing
git checkout nmf_missing
```

TomDLT committed:
Merge branch 'main' into nmf_missing

TomDLT committed:
fix lint

sammymans commented:
okay awesome thanks for getting back to me. Just to confirm, do i just need to change the solver to 'mu' to be able to run with the NaN values? Also does this work with sparse matricies?

PolarBean commented:
This is great! incredibly useful for my project, thank you for developing it. One thing I have noticed is that running NMF with mu solver is much faster when nan values have been replaced by zeroes as compared to running with the nan values left in. Is this expected behaviour? Thank you in advance.

PolarBean commented:
Ah it seems like this is not yet implemented for the partial_fit method of minibatch NMF

PolarBean commented:
The below implementation of _safe_dot may be useful as the numpy implementation of operations on boolean matrices is not highly optimised. in my understanding this means that the majority of the computation in np.ma.dot is computing the mask of the result. since you don't use a masked array as the return value this calculating the return mask is not necessary. So then should give the same result and be much faster. I may be mistaken, but in my tests NMF appears to work well with this. Am I wrong in thinking this will result in an identical model?
```python
def _safe_dot(X, Ht):
    if isinstance(X, np.ma.masked_array):
        X_array = X.data
        X_array[X.mask] = 0
        X = X_array
    if isinstance(Ht, np.ma.masked_array):
        Ht_array = Ht.data
        Ht_array[Ht.mask] = 0
        Ht = Ht_array
    return safe_sparse_dot(X, Ht)



```

PolarBean commented:
The below implementation of _safe_dot may be useful as the numpy implementation of operations on boolean matrices is not highly optimised. in my understanding this means that the majority of the computation in np.ma.dot is computing the mask of the result. since you don't use a masked array as the return value this calculating the return mask is not necessary. So then this should give the same result and be much faster. I may be mistaken, but in my tests NMF appears to work well with this. Am I wrong in thinking this will result in an identical model?
```python
def _safe_dot(X, Ht):
    if isinstance(X, np.ma.masked_array):
        X_array = X.data.copy()
        X_array[X.mask] = 0
        X = X_array
    if isinstance(Ht, np.ma.masked_array):
        Ht_array = Ht.data.copy()
        Ht_array[Ht.mask] = 0
        Ht = Ht_array
    return safe_sparse_dot(X, Ht)



```

jjerphan commented:
@PolarBean : are you interested in pursuing this work? :)

PolarBean commented:
Ah Im currently tied up with work but if there are some small things that are needed to get this merged just let me know!

liaochen1988 commented:
I know this is an old topic and I only found it today. I tried to install the scikit-learn version from pull 8474 but it gives me the following error message

Installed /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/scikit_learn-1.6.dev0-py3.9-macosx-10.9-universal2.egg
Processing dependencies for scikit-learn==1.6.dev0
Searching for scikit-learn==1.6.dev0
Reading https://pypi.org/simple/scikit-learn/
No local packages or working download links found for scikit-learn==1.6.dev0
error: Could not find suitable distribution for Requirement.parse('scikit-learn==1.6.dev0')

Maybe I did not do it correctly. May I know the correct way of using the NMF function that can handle missing values? This function is critical to my project. Thank you in advance!

jjerphan commented:
Hi @liaochen1988 

Can you open a separate issue with a reproducer? 🙂

Thank you!

03bennej commented:
Just wondering if there's any plan to merge this PR in the near future?
